linkerd/stack/src/gate.rs

```diff
@@ -0,0 +1,261 @@
+use crate::Service;
+use futures::{ready, FutureExt};
+use std::{
+    sync::{atomic::AtomicBool, Arc},
+    task::{Context, Poll},
+};
+use tokio::sync::Notify;
+use tokio_util::sync::ReusableBoxFuture;
+use tracing::{debug, trace};
+
+/// A middleware that alters its readiness state according to a gate channel.
+pub struct Gate<S> {
+    inner: S,
+    rx: Rx,
+    is_waiting: bool,
+    waiting: ReusableBoxFuture<'static, ()>,
+}
+
+/// Observes gate state changes.
+#[derive(Clone, Debug)]
+pub struct Rx(Arc<Shared>, Arc<()>);
+
+/// Changes the gate state.
+#[derive(Clone, Debug)]
+pub struct Tx(Arc<Shared>);
+
+#[derive(Debug)]
+struct Shared {
+    open: AtomicBool,
+    notify: Notify,
+    closed: Notify,
+}
+
+/// Creates a new gate channel.
+pub fn channel() -> (Tx, Rx) {
+    let shared = Arc::new(Shared {
+        open: AtomicBool::new(true),
+        notify: Notify::new(),
+        closed: Notify::new(),
+    });
+    (Tx(shared.clone()), Rx(shared, Arc::new(())))
+}
+
+// === impl Rx ===
+
+impl Rx {
+    /// Indicates whether the gate is open.
+    pub fn is_open(&self) -> bool {
+        self.0.open.load(std::sync::atomic::Ordering::Acquire)
+    }
+
+    /// Indicates whether the gate is closed.
+    #[inline]
+    pub fn is_shut(&self) -> bool {
+        !self.is_open()
+    }
+
+    /// Waits for the gate state to change.
+    pub async fn changed(&self) -> bool {
+        self.0.notify.notified().await;
+        self.is_open()
+    }
+
+    /// Waits for the gate state to change to be open.
+    pub async fn opened(&self) {
+        if self.is_open() {
+            return;
+        }
+
+        while !self.changed().await {}
+    }
+}
+
+impl Drop for Rx {
+    fn drop(&mut self) {
+        let Rx(_, handle) = self;
+        if Arc::strong_count(handle) == 1 {
+            self.0.closed.notify_waiters();
+        }
+    }
+}
```

Comment:
take it or leave it: it occurrs to me that if, rather than two `Notify`s and an `Arc<()>`, we used a single `watch<bool>` channel we could get this behavior for free --- to notify the gate, a sender would send a `true` or `false` to indicate closure, and when all the senders are dropped, the watch receiver's `closed` future would complete.

i think the current approach works, too, but we could maybe simplify the implementation a bit this way?

Reply:
`watch::Sender` isn't `Clone`. Our `Tx` is.